File: yasmin/yasmin/concurrence.py
# Copyright (C) 2025 Georgia Tech Research Institute
# Supported by USDA-NIFA CSIAPP Grant. No. 2023-70442-39232
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from typing import Dict, Optional
from threading import Thread, Lock

import yasmin
from yasmin import State
from yasmin.blackboard import Blackboard


class Concurrence(State):
    """
    Runs a series of states in parallel

    The Concurrence class runs a set of states concurrently, waiting
    for the termination of each, and then returns a single output
    according to a provided rule map, or a default outcome if no rule is
    satisfied.

    Attributes:
        _states (Dict[str, State]): The states to run concurrently (name -> state).
        _default_outcome (str): Default outcome.
        _outcome_map (Dict[str, Dict[str, str]]): Specifies which combination of state outputs should produce a given
                                                  overall output.
        _intermediate_outcomes (Dict[str, Optional[str]]): Stores the intermediate outcomes of the concurrent states.
        _mutex (Lock): Mutex for intermediate outcome map.
    """

    def __init__(
        self,
        states: Dict[str, State],
        default_outcome: str,
        outcome_map: Dict[str, Dict[State, str]] = {},
    ) -> None:
        """
        Constructs a State with a set of possible outcomes.

        Args:
            states (Dict[str, State]): A map of state names to states that will run concurrently.
            default_outcome (str): The default outcome to return if no outcome map
            rules are satisfied.
            outcome_map (Dict[str, Dict[str, str]]): A map of outcome names to requirements for achieving
            that outcome.

        Raises:
            ValueError: If either the provided outcomes set or states set are empty.
            ValueError: If the same instance of a state is listed to run concurrently with itself.
            KeyError: If an intermediate outcome is not registered with the correlated state.
        """

        if len(outcome_map.keys()) > 0:
            outcomes = list(set(outcome_map.keys()) | set([default_outcome]))
        else:
            outcomes = [default_outcome]

        super().__init__(outcomes)

        if not states:
            raise ValueError("There must be at least one state")

        # Check for duplicate state instances
        state_instances = list(states.values())
        unique_instances = set(id(state) for state in state_instances)
        if len(unique_instances) != len(state_instances):
            raise ValueError("There are duplicate state names in the states")

        self._states: Dict[str, State] = states
        self._default_outcome: str = default_outcome
        self._outcome_map: Dict[str, Dict[str, str]] = {}

        for outcome, requirements in outcome_map.items():
            self._outcome_map[outcome] = {}
            for state_name, state_outcome in requirements.items():

                if state_name not in self._states:
                    raise KeyError(
                        f"State name '{state_name}' not found in states dictionary"
                    )

                state = self._states[state_name]

                # Check if intermediate outcome belongs to state
                if state_outcome not in state.get_outcomes():
                    raise KeyError(
                        f"Outcome '{state_outcome}' is not registered in state {state_name}"
                    )

                self._outcome_map[outcome][state_name] = state_outcome

        self._intermediate_outcomes: Dict[str, Optional[str]] = {}
        for state_name in self._states.keys():
            self._intermediate_outcomes[state_name] = None

        self._mutex = Lock()

    def execute(self, blackboard: Blackboard) -> str:
        """
        Executes the state's specific logic.

        Args:
            blackboard (Blackboard): A shared pointer to the Blackboard to use during
            execution.

        Returns:
            str: A string representing the outcome of the execution.
        """

        state_threads = []

        for state_name, state in self._states.items():
            state_threads.append(
                Thread(
                    target=self.execute_and_save_state,
                    args=(
                        state,
                        state_name,
                        blackboard,
                    ),
                )
            )
            state_threads[-1].start()

        for t in state_threads:
            t.join()

        satisfied_outcomes = []
        for _, (outcome, requirements) in enumerate(self._outcome_map.items()):
            satisfied = True

            for state_name, state_outcome in requirements.items():
                satisfied = satisfied and (
                    self._intermediate_outcomes[state_name] == state_outcome
                )

            if satisfied:
                satisfied_outcomes.append(outcome)

        if len(satisfied_outcomes) == 0:
            return self._default_outcome

        if len(satisfied_outcomes) > 1:
            yasmin.YASMIN_LOG_WARN(
                f"More than one satisfied outcome after concurrent state execution."
            )

        return satisfied_outcomes[0]

    def execute_and_save_state(
        self, state: State, state_name: str, blackboard: Blackboard
    ) -> None:
        """
        Executes a state and saves its outcome to the intermediate map.

        Args:
            state (State): A state to execute.
            state_name (str): The name of the state to execute.
            blackboard (Blackboard): An instance of Blackboard that provides the context for execution.
        """
        outcome = state(blackboard)
        with self._mutex:
            self._intermediate_outcomes[state_name] = outcome

    def get_states(self) -> Dict[str, State]:
        """
        Returns the map of states managed by this concurrence state.

        Returns:
            Dict[str, State]: A map of state names to states.
        """
        return self._states

    def get_outcome_map(self) -> Dict[str, Dict[str, str]]:
        """
        Returns the outcome map for this concurrence state.

        Returns:
            Dict[str, Dict[str, str]]: A map of outcome names to their requirements.
        """
        return self._outcome_map

    def get_default_outcome(self) -> str:
        """
        Returns the default outcome for this concurrence state.

        Returns:
            str: The default outcome as a string.
        """
        return self._default_outcome

    def cancel_state(self) -> None:
        """
        Cancels the current state execution.

        This method sets the canceled flag to true and logs the action.
        """
        for key in self._states:
            self._states[key].cancel_state()

        super().cancel_state()

    def __str__(self) -> str:
        """
        Converts the state to a string representation.

        Returns:
            str: A string representation of the state.
        """
        result = "Concurrence ["

        for i, key in enumerate(self._states):
            result += f"{key} ({self._states[key]})"

            if i < len(self._states) - 1:
                result += ", "

        result += "]"
        return result

File: yasmin/src/yasmin/concurrence.cpp
// Copyright (C) 2025 Georgia Tech Research Institute
// Supported by USDA-NIFA CSIAPP Grant. No. 2023-70442-39232
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "yasmin/concurrence.hpp"

#include <algorithm>
#include <exception>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

#include "yasmin/logs.hpp"
#include "yasmin/orthogonal_state.hpp"
#include "yasmin/state.hpp"
#include "yasmin/state_machine.hpp"
#include "yasmin/state_utils.hpp"
#include "yasmin/types.hpp"

using namespace yasmin;

Concurrence::Concurrence(const StateMap &states,
                         const std::string &default_outcome,
                         const OutcomeMap &outcome_map,
                         const ParameterMappingsMap &parameter_mappings)
    : State(yasmin::generate_possible_outcomes(outcome_map, default_outcome)),
      states(states), default_outcome(default_outcome),
      outcome_map(outcome_map), parameter_mappings(parameter_mappings) {

  // Require at least one state
  if (states.empty()) {
    throw std::invalid_argument("Map of concurrent states cannot be empty");
  }

  // Check for duplicate state instances
  std::set<std::shared_ptr<State>> unique_instances;
  for (const auto &[state_name, state] : states) {
    if (unique_instances.find(state) != unique_instances.end()) {
      throw std::invalid_argument(
          "There are duplicate state instances in the states");
    }
    unique_instances.insert(state);
  }

  // Validate outcome map and prepare intermediate outcomes map
  for (const auto &[outcome, requirements] : outcome_map) {
    if (requirements.empty()) {
      throw std::invalid_argument(
          "Outcome '" + outcome +
          "' in the outcome map must have at least one requirement");
    }
    for (const auto &[state_name, intermediate_outcome] : requirements) {
      // Check if state name exists in the states map
      auto state_it = states.find(state_name);
      if (state_it == states.end()) {
        throw std::invalid_argument(
            "State name '" + state_name +
            "' in the outcome map under outcome '" + outcome +
            "' is not provided in the concurrent set of states to run");
      }

      // Check if intermediate outcome is valid for the state
      State::SharedPtr state = state_it->second;
      if (state->get_outcomes().find(intermediate_outcome) ==
          state->get_outcomes().end()) {
        throw std::invalid_argument(
            "Intermediate outcome '" + intermediate_outcome +
            "' under outcome '" + outcome +
            "' of the outcome map is not a valid outcome of state '" +
            state->to_string() + "'");
      }

      this->intermediate_outcome_map.insert({state_name, ""});
    }
  }
}

void Concurrence::set_parameter_mappings(
    const std::string &state_name,
    const ParameterMappings &parameter_mappings) {
  if (this->states.find(state_name) == this->states.end()) {
    throw std::invalid_argument(
        "State name '" + state_name +
        "' is not provided in the concurrent set of states to run");
  }

  this->parameter_mappings[state_name] = parameter_mappings;
  this->configured.store(false);
}

const ParameterMappingsMap &
Concurrence::get_parameter_mappings() const noexcept {
  return this->parameter_mappings;
}

void Concurrence::apply_parameter_mappings(const std::string &state_name,
                                           const State::SharedPtr &state) {
  yasmin::apply_parameter_mappings("Concurrence", this->parameter_mappings,
                                   state_name, *this, state);
}

void Concurrence::configure() {
  if (yasmin::check_already_configured(this->configured, "Concurrence",
                                       this->to_string().c_str()))
    return;

  for (const auto &[state_name, state] : this->states) {
    this->apply_parameter_mappings(state_name, state);
    state->configure();
  }

  this->configured.store(true);
}

std::string Concurrence::execute(Blackboard::SharedPtr blackboard) {
  this->configure();
  std::vector<std::thread> state_threads;

  // Reset stored intermediate outcomes before starting a new execution.
  for (auto &[state_name, intermediate_outcome] :
       this->intermediate_outcome_map) {
    (void)state_name;
    intermediate_outcome.clear();
  }

  // Initialize the parallel execution of all the states.
  // Each branch receives a blackboard copy that shares the underlying storage
  // but keeps an isolated remapping scope.
  for (const auto &[state_name, state] : this->states) {
    Blackboard::SharedPtr thread_blackboard =
        std::make_shared<Blackboard>(*blackboard);
    state_threads.push_back(std::thread([this, state_name, state,
                                         thread_blackboard]() {
      std::string outcome = (*state.get())(thread_blackboard);
      const std::lock_guard<std::mutex> lock(this->intermediate_outcome_mutex);
      this->intermediate_outcome_map[state_name] = outcome;
    }));
  }

  // Wait for states to finish
  for (std::thread &state_thread : state_threads) {
    if (state_thread.joinable()) {
      state_thread.join();
    }
  }

  // Handle a cancel
  if (this->is_canceled()) {
    return this->default_outcome;
  }

  // Build final outcome
  Outcomes satisfied_outcomes = yasmin::evaluate_satisfied_outcomes(
      this->outcome_map,
      [this](const std::string &state_name) -> const std::string & {
        auto it = this->intermediate_outcome_map.find(state_name);
        if (it == this->intermediate_outcome_map.end() || it->second.empty()) {
          throw std::runtime_error("An intermediate outcome for state '" +
                                   state_name + "' was not received.");
        }
        return it->second;
      });

  return yasmin::resolve_outcome(satisfied_outcomes, this->default_outcome,
                                 "concurrent state execution (" +
                                     this->to_string() + ")");
}

void Concurrence::cancel_state() {
  for (const auto &[state_name, state] : this->states) {
    state->cancel_state();
  }
  yasmin::State::cancel_state();
}

const StateMap &Concurrence::get_states() const noexcept {
  return this->states;
}

const OutcomeMap &Concurrence::get_outcome_map() const noexcept {
  return this->outcome_map;
}

const std::string &Concurrence::get_default_outcome() const noexcept {
  return this->default_outcome;
}

void Concurrence::validate(bool strict_mode) {
  for (const auto &[state_name, state] : this->states) {
    (void)state_name;
    // Use get_inner_state() to unwrap proxy wrappers (e.g. PythonStateHolder)
    // so that Python-loaded states are correctly identified.
    auto *inner = state->get_inner_state();
    if (auto *sm = dynamic_cast<StateMachine *>(inner)) {
      sm->validate(strict_mode);
    } else if (auto *c = dynamic_cast<Concurrence *>(inner)) {
      c->validate(strict_mode);
    } else if (auto *o = dynamic_cast<OrthogonalState *>(inner)) {
      o->validate(strict_mode);
    }
  }
}

std::string Concurrence::to_string() const {
  return "Concurrence [" +
         yasmin::join(this->states, ", ",
                      [](const auto &p) {
                        return p.first + " (" + p.second->to_string() + ")";
                      }) +
         "]";
}
